File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2023/Y23D05.kt
package org.gristle.pdxpuzzles.advent.y2023

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.parsing.blankSplit
import org.gristle.pdxpuzzles.utilities.parsing.getLongList
import kotlin.math.min

class Y23D05(input: String) : Day {

    data class Listing(val sourceStart: Long, val offset: Long, val length: Long) {
        val sourceEnd = sourceStart + length - 1
    }

    private val seeds: List<Long>
    private val conversions: List<List<Listing>>

    init {
        val stanzas = input.blankSplit().map { it.getLongList() }
        seeds = stanzas[0]
        conversions = stanzas
            .drop(1)
            .map { mapNumbers ->
                mapNumbers
                    .chunked(3) { (destinationStart, sourceStart, length) ->
                        Listing(sourceStart, destinationStart - sourceStart, length)
                    }.sortedBy(Listing::sourceStart)
            }
    }

    private fun solve(seedRanges: List<LongRange>): Long = seedRanges
        .minOf { seedRange -> // for each seedRange, find the smallest end number
            // feed a list of ranges through a gauntlet of conversion stages. the list of ranges will grow each step
            // as each conversion handles different parts of the range differently.
            val subRanges = conversions.fold(listOf(seedRange)) { ranges, listings ->
                ranges.flatMap { range ->
                    buildList {
                        // go through each listing in ascending order, adding subRanges where appropriate, mapping to
                        // destination where appropriate. In order to avoid using a mutable "last" variable, we
                        // put everything in a fold that updates an internal "next" and ultimately outputs a "last"
                        // value to use to complete the subRanges once we've run through all the listings.
                        val last = listings.fold(range.first) { next, listing ->
                            if (range.last >= listing.sourceStart && next <= listing.sourceEnd) {
                                if (next < listing.sourceStart) {
                                    add(next until listing.sourceStart)
                                }
                                val mapEnd = min(range.last, listing.sourceEnd)
                                add(next + listing.offset..mapEnd + listing.offset)
                                mapEnd + 1
                            } else {
                                next
                            }
                        }
                        // clean up subRanges
                        if (last <= range.last) add(last..range.last)
                    }
                }
            }
            // we only care about the lowest number of all these subRanges
            subRanges.minOf { it.first }
        }

    override fun part1() = solve(seeds.map { it..it })

    override fun part2() = solve(seeds.chunked(2) { (start, length) -> start until start + length })
}

fun main() = Day.runDay(Y23D05::class)

//    Class creation: 11ms
//    Part 1: 379811651 (4ms)
//    Part 2: 27992443 (4ms)
//    Total time: 20ms

@Suppress("unused")
private val sampleInput = listOf(
    """seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4""",
)

File: rust/advent/src/bin/2023/y23d05.rs
use std::cmp::min;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::{
    parsing::get_numbers::get_numbers,
    structs::stopwatch::{ReportDuration, Stopwatch},
};

type Input = (Vec<i64>, Vec<Vec<Listing>>);
type Output = i64;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(23, 5).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

#[derive(Clone, Copy, Debug)]
struct Listing {
    source_start: i64,
    offset: i64,
    len: i64,
}

impl Listing {
    fn source_end(&self) -> i64 {
        self.source_start + self.len - 1
    }
}

fn parse_input(input: &str) -> Input {
    let stanzas: Vec<Vec<i64>> = input.split("\n\n").map(get_numbers).collect();

    let seeds = stanzas[0].to_owned();
    let conversions = stanzas[1..]
        .iter()
        .map(|map_numbers| {
            map_numbers
                .iter()
                .tuples()
                .map(|(destination_start, source_start, length)| Listing {
                    source_start: *source_start,
                    offset: destination_start - source_start,
                    len: *length,
                })
                .sorted_by_key(|listing| listing.source_start)
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();
    (seeds, conversions)
}

fn solve(conversions: &[Vec<Listing>], seed_ranges: Vec<(i64, i64)>) -> Output {
    seed_ranges
        .iter()
        .map(|seed_range| {
            let sub_ranges: Vec<(i64, i64)> =
                conversions
                    .iter()
                    .fold(vec![seed_range.to_owned()], |ranges, listings| {
                        let test: Vec<(i64, i64)> = ranges
                            .iter()
                            .flat_map(|range| {
                                let mut sub_ranges: Vec<(i64, i64)> = Vec::new();
                                let (range_first, range_last) = range;
                                let last = listings.iter().fold(*range_first, |next, listing| {
                                    if range_last >= &listing.source_start
                                        && next <= listing.source_end()
                                    {
                                        if next < listing.source_start {
                                            sub_ranges
                                                .push((next.to_owned(), listing.source_start - 1));
                                        }
                                        let se = listing.source_end();
                                        let map_end = min(range_last, &se);
                                        sub_ranges.push((
                                            next + listing.offset,
                                            map_end + listing.offset,
                                        ));
                                        *map_end + 1
                                    } else {
                                        next
                                    }
                                });
                                if &last <= range_last {
                                    sub_ranges.push((last.to_owned(), range_last.to_owned()));
                                }
                                sub_ranges
                            })
                            .collect();
                        test
                    });

            sub_ranges
                .iter()
                .map(|(first, _)| first.to_owned())
                .min()
                .unwrap()
        })
        .min()
        .unwrap()
}

fn part1(input: &Input) -> Output {
    let (seeds, conversions) = input;
    let seed_ranges: Vec<(i64, i64)> = seeds.iter().map(|seed| (*seed, *seed)).collect();
    solve(conversions, seed_ranges)
}

fn part2(input: &Input) -> Output {
    let (seeds, conversions) = input;
    let seed_ranges: Vec<(i64, i64)> = seeds
        .chunks(2)
        .filter_map(|chunk| {
            if let [start, length] = chunk {
                Some((*start, start + length - 1))
            } else {
                None
            }
        })
        .collect();
    solve(conversions, seed_ranges)
}

#[test]
fn default() {
    let input = get_input(23, 5).unwrap();
    let input = parse_input(&input);
    assert_eq!(379811651, part1(&input));
    assert_eq!(27992443, part2(&input));
}

// Input parsed (50μs)
// 1. 379811651 (12μs)
// 2. 27992443 (37μs)
// Total: 102μs